src/ray/util/event.cc

```diff
@@ -16,11 +16,215 @@
+    FlushLocked();
+    if (!shutdown_) {
+      timer_.expires_at(timer_.expiry() +
+                        boost::asio::chrono::seconds(loop_flush_wait_second_));
+      timer_.async_wait(boost::bind(&LogFileObject::LoopFlush, this));
+    }
+  }
+}
+
+bool LogFileObject::CreateLogfile() {
+  file_ = fopen(GetFileNameByIndex().c_str(), "a+");
+  if (file_ == NULL) {
+    return false;
+  }
+  return true;
+}
+
+std::string LogFileObject::GetFileNameByIndex(int index) {
+  std::string result = log_dir_ + file_name_;
+  if (index == 0) {
+    result += ".log";
+    return result;
+  } else {
+    result += "." + std::to_string(index) + ".log";
+    return result;
+  }
+}
+
+void LogFileObject::Write(const char *message, int message_len, bool force_flush) {
+  absl::MutexLock lock(&mutex_);
+  if ((file_length_ >> 20) >= max_log_size_) {
+    if (file_ != NULL) {
+      fclose(file_);
+      // Rename file when the file size is excess max_log_size_
+      rotate_index_++;
+      rename(GetFileNameByIndex().c_str(), GetFileNameByIndex(rotate_index_).c_str());
+      if (rotate_index_ > max_rotate_num_) {
+        for (int i = 1; i <= max_rotate_num_; ++i) {
+          rename(GetFileNameByIndex(i + 1).c_str(), GetFileNameByIndex(i).c_str());
+        }
+        rotate_index_ = max_rotate_num_;
+      }
+    }
+    file_ = NULL;
+    file_length_ = bytes_since_flush_ = 0;
+    rollover_attempt_ = rollover_attempt_frequency_ - 1;
+  }
+
+  if (file_ == NULL) {
+    if (++rollover_attempt_ != rollover_attempt_frequency_) return;
+    rollover_attempt_ = 0;
+    if (!CreateLogfile()) {
+      fprintf(stderr, "COULD NOT CREATE LOGFILE '%s'!\n", file_name_.c_str());
+      return;
+    }
+  }
+
+  if (!stop_writing_) {
+    errno = 0;
+    // use message_len + 1 because LogSink the last char is '\n' "
+    fwrite(message, 1, message_len, file_);
+    if (errno == ENOSPC) {
+      stop_writing_ = true;
+      return;
+    } else {
+      file_length_ += message_len;
+      bytes_since_flush_ += message_len;
+    }
+  }
+
+  if (force_flush || bytes_since_flush_ > max_flush_size_) {
+    FlushUnlocked();
+  }
+}
+
+///
+/// LogEventReporter
+///
+LogEventReporter::LogEventReporter(rpc::Event_SourceType source_type, std::string log_dir)
+    : log_dir_(log_dir) {
+  RAY_CHECK(log_dir_ != "");
+  if (log_dir_.back() != '/') {
+    log_dir_ += '/';
+  }
+  boost::filesystem::create_directories(log_dir_);
+  file_name_ = "event_" + Event_SourceType_Name(source_type);
+  if (source_type == rpc::Event_SourceType::Event_SourceType_CORE_WORKER ||
+      source_type == rpc::Event_SourceType::Event_SourceType_COMMON) {
+    file_name_ += "_" + std::to_string(getpid());
+  }
+
+  log_sink_ = std::unique_ptr<LogFileObject>(new LogFileObject(log_dir_, file_name_));
+}
+
+std::string LogEventReporter::EventToString(const rpc::Event &event) {
+  std::stringstream result;
+
+  result << "{\"time_stamp\":\"" << event.timestamp() << "\",\"severity\":\""
```

Comment:
boost::property_tree::json_parser::write_json could be better if we use it here

Reply:
I will try to use this.